### PR TITLE
Added telegram.py as reporter

### PR DESCRIPTION
**Pull request by denisgolius**
For using add this lines yo master.cfg and enter your variables for TOKEN & @name_bot:

from telegram_reporter import TelegramStatus
c['services'].append(TelegramStatus(auth_token="TOKEN", chat_id="@name_bot"))

Sorry, but some tests from developers link was failed.. also I don't know how to write auto tests...yet... 😕

### CODE REVIEW

**Comment by tardyp** [comment]
Hi,
Please have a look at our developer quickstart to know how to run and develop the tests.

http://docs.buildbot.net/latest/developer/quickstart.html

A test for a similar reporter can be found here:
https://github.com/buildbot/buildbot/blob/b7c8e76d5c89e08186e5100c24111e9b3a9a8035/master/buildbot/test/unit/test_reporters_hipchat.py#L15

File: master/buildbot/reporters/telegram.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            'new': 'Buildbot started build %s here: %s' % (build['builder']['name'], build['url']),
+            'finished': 'Buildbot finished build %s with result %s here: %s'
+                        % (build['builder']['name'], statusToString(build['results']), build['url'])
+        }
+        return event_messages.get(event_name, '')
+
+    def send(self, build, key):
+        data = self.getMessage(build, key)
+        message = "{0}".format(data)
+        self.bot.send_message(chat_id=self.chat_id, text=message)
```
Comment (tardyp) [inline master/buildbot/reporters/telegram.py]: This needs to be done in a thread. Telegram is a synchronous API, and thus during the time, the network packets are going around the globe, buildbot cannot do anything else.

File: master/buildbot/reporters/telegram.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def buildStarted(self, key, build):
+        return self.getMoreInfoAndSend(build, key[2])
+
+    def buildFinished(self, key, build):
+        return self.getMoreInfoAndSend(build, key[2])
+
+    @defer.inlineCallbacks
+    def getMoreInfoAndSend(self, build, key):
+        yield getDetailsForBuild(self.master, build)
+        yield self.send(build, key)
```
Comment (tardyp) [inline master/buildbot/reporters/telegram.py]:
```suggestion
      yield threads.deferToThread(self._thd_send, build, key)
```

File: master/buildbot/reporters/telegram.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    def getMessage(self, build, event_name):
+        event_messages = {
+            'new': 'Buildbot started build %s here: %s' % (build['builder']['name'], build['url']),
+            'finished': 'Buildbot finished build %s with result %s here: %s'
+                        % (build['builder']['name'], statusToString(build['results']), build['url'])
+        }
+        return event_messages.get(event_name, '')
+
+    def send(self, build, key):
```
Comment (tardyp) [inline master/buildbot/reporters/telegram.py]: make it explicit that this function is running from a worker thread
```suggestion
    def _thd_send(self, build, key):
```

File: master/buildbot/reporters/telegram.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    def buildFinished(self, key, build):
+        return self.getMoreInfoAndSend(build, key[2])
+
+    @defer.inlineCallbacks
+    def getMoreInfoAndSend(self, build, key):
+        yield getDetailsForBuild(self.master, build)
+        yield self.send(build, key)
+
+    def getMessage(self, build, event_name):
+        event_messages = {
```
Comment (tardyp) [inline master/buildbot/reporters/telegram.py]: This is not a particularly useful style, as this is formatting both messages all the time.
Please stick to good old if .. else

**Comment by tardyp** [comment]
Also needs docs and tests (MR description is not doc :) )
and news fragments.

Please read our contribution guide. This is important in order to avoid loss your time and ours.
http://docs.buildbot.net/latest/developer/pull-request.html

File: master/buildbot/reporters/telegram.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+from __future__ import absolute_import
```
Comment (sa2ajj) [inline master/buildbot/reporters/telegram.py]: do we need these two? master [expects](http://docs.buildbot.net/current/manual/installation/requirements.html) a reasonable version of Python 3.

File: master/buildbot/reporters/telegram.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+from __future__ import print_function
+
+from future.utils import string_types
```
Comment (sa2ajj) [inline master/buildbot/reporters/telegram.py]: ditto

File: master/buildbot/reporters/telegram.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    def checkConfig(self, auth_token, chat_id, **kwargs):
+        if not isinstance(auth_token, string_types):
+            config.error('auth_token must be a string')
+
+    @defer.inlineCallbacks
+    def reconfigService(self, auth_token, chat_id, **kwargs):
+        yield BuildbotService.reconfigService(self, **kwargs)
+
+        self.auth_token = auth_token
```
Comment (sa2ajj) [inline master/buildbot/reporters/telegram.py]: are all these data member part of public API? if not, please prepend them with `_`, like `self._auth_token`

File: master/buildbot/reporters/telegram.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        event_messages = {
+            'new': 'Buildbot started build %s here: %s' % (build['builder']['name'], build['url']),
+            'finished': 'Buildbot finished build %s with result %s here: %s'
+                        % (build['builder']['name'], statusToString(build['results']), build['url'])
+        }
+        return event_messages.get(event_name, '')
+
+    def send(self, build, key):
+        data = self.getMessage(build, key)
+        message = "{0}".format(data)
```
Comment (sa2ajj) [inline master/buildbot/reporters/telegram.py]: several questions:
* is it really needed to have a separate `message`?
* why not use `str(data)`?

**Comment by p12tic** [comment]
Just in case someone wants to resurrect the PR later.

**Pull request closed without merging**